### ISSUE TITLE
[ENH] should we adopt defensive upper version bounds?

### ISSUE THREAD

**Issue by fkiraly** [alan-turing-institute/sktime#1480]
The `statsmodels` breakage has again shown us that under the current model, `sktime` is vulnerable to interface changes in other packages at version zero.

Should we perhaps adopt a more defensive stance?

More precisely:

* upper version bounds for *all* dependencies
* at each release, we bump the upper bound to the current versions, or the highest versions that don't cause breakage
* if at a release we cannot move the upper bound to current version, we open a maintenance issue to log this

Opinions, @mloning, @aiwalter, @freddyaboulton?

**Comment by fkiraly** [comment]
One more data point - we had a near miss with `pandas` 1.5.0 that would have shot apart one of our most important estimators, `Lag`, plus a few other smaller issues, related to `freq` stuff. It looks like this came without clear warning.

Because we had a defensive version bound `<1.5.0`, this did not propagate through to `sktime` in deployment.

To me, looks like a point in favour.

I would suggest a policy where we keep defensive version bounds as long as we can react within a few weeks. As long as testing indicates, for a given dependency, that `sktime` would have broken, we should keep those bounds (say, until three times in a row of MINOR updates without issues?)

**Comment by fkiraly** [comment]
@lmmentel, I wanted to show someone the discussion we had on this, e.g., that adopting too strict devensive bounds or even pins may lead to packages getting out of sync or resulting in contradictory requirements.

Not sure where that was - would appreciate a pointer, or a summary of your opinion here, in case it was only on slack.

**Comment by fkiraly** [comment]
@achieveordie in #3816:

> Of course, one can argue that our build should hold more specific versions as said by @RikStarmans, but I'm not sure what would be the "right" approach. On one hand, the current code enables us to catch and correct such changes while on the other we'd ensure the test suite works without fail.

These considerations, and earlier instances of "update results in failure" have made me open this issue, and prompted us to introduce some defensive version bounds on core packages.

The situation and trade-off is, however, not as straightforward as it may seem.
Unfortunately it may be that the discussion we had internally was not recorded properly, or I'm just failing at finding it.

The main counterargument is as above: you introduce too many pins or bounds, then it becomes more and more difficult to find *any* working configuration if you are installing many packages - the argument is by the Kantian imperative, "suppose everyone was doing this, what would happen"?

You could now argue, ok, frameworks are more "top-level" than component packages, but people can also opt to install multiple frameworks, which may have incompatible dependency bounds.

**Comment by fkiraly** [comment]
(It really looks like I'm having a monologue with myself here, but there was a very heated discussion on our internal slack and in dev meetings, as said, it may not have been recorded)

**Comment by yarnabrina** [comment]
I agree with all of these in principle, but would like to suggest slight modifications to consider as well for being specific:

* **both** upper and lower bounds for **all** dependencies
  * initiate lower bounds as tested (current as of now unless someone manually tests for lower versions) minor version (assuming major.minor.patch format)
  * initiate upper bounds as next non-breaking version (next major if major is non-zero, next minor if major is zero - this is probably standard)
* regularly update bounds
  * upper bounds are automatically updated by dependabot
  * lower bounds are updated if there are at least k (2/3 - needs to be decided) minor releases have been done since

Let me know what do you think.

(Originally posted in https://github.com/sktime/sktime/pull/5136#issuecomment-1688398835)

**Comment by fkiraly** [comment]
Finally this has thread has moved on from a monologue to a real discussion 😃
Thanks, @yarnabrina!

> both upper and lower bounds for all dependencies 

Agreed as detailed, assuming we do not change upper bounds as current.

That is because some upper bounds are on minor, even though major is non-zero. In all cases I can recall, this is because there have been breakages at minor versions repeatedly - e.g., `pandas`.

> regularly update bounds 

Agreed with handling upper.

Disagreed with handling lower. Why should we move these if nothing breaks?

---

Adding links to related:

* @romanlutz suggested lower bounds and testing as well, in #5081
* PR to introduce testing of lower bounds - https://github.com/sktime/sktime/pull/4016 - fails with strange `numpy` compilation error

**Comment by yarnabrina** [comment]
> My high-level stance would be, being defensive-reactive with the major/minor distinction. If minor breaks at multiple updates, move it to minor or leave it there. If minor upgrade goes fine for k updates in a row, move it to major.

Makes sense. We can always keep it minor even to be totally safe.

> Disagreed with handling lower. Why should we move these if nothing breaks?

I am suggesting this so that our bounds are tested, at least recently. CI only tests latest versions, and if we never update lower bound, we have no way to know for sure if it'd work or not. But from user's point of view, `sktime` is saying it supports this version specifically but actually it's not.

For example, suppose we add a new bound now `pkg>1.2,<1.3`. This is currently tested, and is good. Also assume, the package make few more minor releases over time, and each time with new `dependabot` PR, we have reached `pkg>1.2,<1.8`. At this point, we are testing only `1.7` and nothing else. And developers may add features on the estimator with `pkg` dependency that actually uses `>1.4` features and it passes all tests, as there's no easy way to test for all lower versions.

Regularly bumping lower bounds will reduce this chance I think, as there were probably never any breaking change in `pkg`, but over time small changes had merged together into one.

(The context is very very similar, if not identical to `statsforecast` PR on `levels`.)

---

I went through a little bit of `fairlearn` repository, and their number of dependencies is just 4 (`numpy`, `pandas`, `scikit-learn` and `scipy` - very closely related/user together). I am a bit pessimistic to hope that it'd be easy/possible to translate their lower bound testing pattern into `sktime`'s with over 30 soft dependencies on top of 6 requirements.

**Comment by fkiraly** [comment]
> I am suggesting this so that our bounds are tested, at least recently. CI only tests latest versions, and if we never update lower bound, we have no way to know for sure if it'd work or not. 

I see - my counterproposal would be to test the minimal lower bounds, that ensures we test them. That is, fix https://github.com/sktime/sktime/pull/4016 and run it with the CI.

**Comment by fkiraly** [comment]
> Regularly bumping lower bounds will reduce this chance I think

Imo the strongest counterargument is, deployment setups may have old versions of some packages - bumping lower versions is going to make having a valid setup more difficult to users who have a restricted deployment environment.

**Comment by yarnabrina** [comment]
Sorry for late reply, but my disagreement to keep lower bounds unchanged or running lowest configurations still remains. I'll try to present my point of view below.

Suppose this is an extra:

```toml
extra_name = [
    "pkg_1>=lb_1,<hb_1",
    "pkg_2>=lb_2,<hb_2",
    ...,
    "pkg_k>=lb_k,<hb_k",
]
```

In this case, over time CI only tests against the combination of the latest `patch` releases for `pkg_1`, `pkg_2`, ..., `pkg_k`, and nothing else. The proposal to test againgst lowest versions only cover the case for one another element in the entire universal set, and nothing more.

But the above extra means (to the package user) that **_any_ release of `pkg_1` between `lb_1` and `hb_1` is compatible to _any_ combination of versions of `pkg_2`, `pkg_3`, ..., `pkg_k`, as long as the versions satify the respective bounds**. This set has a significant cardinality to test in CI for all pull requests (or even just manual releases). We are talking about the every single releases for all soft dependencies of `sktime` (or a particular extra, in case #5136 is approved and merged). Even we restrict ourselves to just `minor` reeases (and not `patch` releases assuming they do not intriduce any breaking changes), I shall argue that it's still not very feasible.

On the other hand, if we go along the continuous update of lower bound, subject to a specific time limit or specific number of new releases since existing lower bound, we are decreasing this chance. I admit that ee'll still not be able to ensure all possibilities, but it's a much better coverage.

(FYI: @fkiraly, @benHeid, @achieveordie )

**Comment by fkiraly** [comment]
Ok, I will try to comment on your argument and try to expand my argument.

> I shall argue that it's still not very feasible.

Regarding your argument: I don't think we *need* to test for all combinatorial combinations here - we can say that we are reasonably sure that if testing works for the extremal combinations, it works in-between. In general it need not be true, but it seems plausible enough.

Compare to the concept of value testing and certificates for edge and generic cases, e.g., in Code Complete (McConnell). Here also we are unable to test all combinatorial combinations, but select a few combination of input cases where the plausibility argument of generalization to all cases is strong.

Now, my counterargument to moving up version bounds:

this is coming from the general deprecation policy

https://www.sktime.net/en/stable/developer_guide/deprecation.html

and the high-level requirement, "don't intentionally break anything for anyone downstream without sufficient prior announcement" (where the announcement is defined in the policy as one or multiple minor cycles with warning in the changed use case). A corollary is not doing this without a really good reason.

Here's how this implies for version bounds: let's say we have a user who has an upper bound that is much lower than `hb_1`, on `pkg_1`, which comes from, say, a secure mirror that lags half a year behind the current state or so. Let's even say it is exactly `lb_1`, and so far everything works.

Now, we bump the lower bound without forewarning, and suddenly that user is no longer able to satisfy the bounds. So they are now stuck with the last version of `sktime` where lower bound is `lb_1`, for a while, in a situation where we wouldn't have needed to make them stuck.
